### DISCUSSION TITLE
Log the origin of each property.

### DISCUSSION THREAD

**Discussion by edward3h** [apple/pkl#604] [Q&A] [answered]
We have an in-house configuration setup. It reads 'ini' files (properties format) in a hierarchy, overwriting earlier property values with later ones. Our Java loader logs what the final source of each property is, to help with debugging configuration.

We also have a combined output file that has a comment with each property to say where it came from.

Is there a way to do something similar in Pkl? Either some logging when my Java application loads config, or generate a file that shows where properties came from?

**Comment by bioball** [comment, marked as answer]
We don't have a feature like this yet. But, provenance is something that we want to improve and may be something we support in a future release.
